### PR TITLE
BUG: Fix busday_count for reversed dates

### LINKED ISSUE #23197: BUG: busday_count returns wrong results with begindates > enddates
### Describe the issue:

According to the documentation, busday_count

> Counts the number of valid days between begindates and enddates, not including the day of enddates. If enddates specifies a date value that is earlier than the corresponding begindates date value, the count will be negative.

However, [in the implementation](https://github.com/numpy/numpy/blob/6dadb8c40451e934075904f6acdfe341d3b8762e/numpy/core/src/multiarray/datetime_busday.c#L384), if begindates is greater than enddates, the dates are swapped and the result is negated. This has the effect of including the enddates and excluding begindates, which contradicts the above spec.

### Reproduce the code example:

```python
from datetime import date
import numpy as np

d1 = date(1999, 12, 30)
d2 = date(2000, 1, 1)
assert np.busday_count(d1, d2) == 2   # 1999-12-30 and 1999-12-31
assert np.busday_count(d2, d1) == -1  # 1999-12-31 only, end date is excluded
```

### Error message:

```shell
Traceback (most recent call last):
  File "/home/elt/code/numpy-test/busday.py", line 8, in <module>
    assert np.busday_count(d2, d1) == -1
AssertionError
```

### Runtime information:

Output from `import sys, numpy; print(numpy.__version__); print(sys.version)`
```shell
1.24.1
3.10.6 (main, Nov 14 2022, 16:10:14) [GCC 11.3.0]
```

Output from `print(numpy.show_runtime())`
```shell
[{'simd_extensions': {'baseline': ['SSE', 'SSE2', 'SSE3'],
                      'found': ['SSSE3',
                                'SSE41',
                                'POPCNT',
                                'SSE42',
                                'AVX',
                                'F16C',
                                'FMA3',
                                'AVX2'],
                      'not_found': ['AVX512F',
                                    'AVX512CD',
                                    'AVX512_KNL',
                                    'AVX512_KNM',
                                    'AVX512_SKX',
                                    'AVX512_CLX',
                                    'AVX512_CNL',
                                    'AVX512_ICL']}},
 {'architecture': 'Haswell',
  'filepath': '/home/elt/.local/lib/python3.10/site-packages/numpy.libs/libopenblas64_p-r0-15028c96.3.21.so',
  'internal_api': 'openblas',
  'num_threads': 16,
  'prefix': 'libopenblas',
  'threading_layer': 'pthreads',
  'user_api': 'blas',
  'version': '0.3.21'}]
```
### Context for the issue:

This breaks the invariant that
```python
count = numpy.busday_count(d1, d2)
numpy.busday_offset(d1, count, "forward" if count > 0 else "backward") == d2
```
for every d1 and every valid d2.

### PR DESCRIPTION
**Pull request by eendebakpt**
Fixes #23197

This is a backwards incompatible change, so we need to add a release note.

### CODE REVIEW

File: numpy/core/src/multiarray/datetime_busday.c [hunk 1]
```diff
@@ -396,7 +401,7 @@ apply_business_day_count(npy_datetime date_begin, npy_datetime date_end,
                                         holidays_begin, holidays_end);
 
     /* Start the count as negative the number of holidays in the range */
-    count = -(holidays_end - holidays_begin);
+    count += -(holidays_end - holidays_begin);
```
Comment (eltoder) [inline numpy/core/src/multiarray/datetime_busday.c:404]: I think this change and the change to initialize count to 0 are not necessary. Just incrementing the dates is enough.

File: doc/release/upcoming_changes/23229.compatibility.rst [hunk 2]
```diff
@@ -0,0 +1,2 @@
+- The ``busday_count`` method now correctly handles cases where the ``begindates`` and later in time
```
Comment (seberg) [inline doc/release/upcoming_changes/23229.compatibility.rst:1]: Could you rephrase or add to say what "correctly" means here?  Something like: "previously the ``enddates`` was included when it should always be excluded." maybe?

"and later" -> is later or so.

Comment (eendebakpt) [reply doc/release/upcoming_changes/23229.compatibility.rst]: Done

**Comment by seberg** [comment]
Hmmm, somehow I had misremembered it as just including the enddate, but it actually does exclude the startdate for negative values (i.e. it always excludes the larger date).  I am OK with giving it a shot, but it does make me a bit less confident that nobody might be working around this in weird ways.

I guess we merge it ping the mailing list?

**Comment by eendebakpt** [comment]
> Hmmm, somehow I had misremembered it as just including the enddate, but it actually does exclude the startdate for negative values (i.e. it always excludes the larger date). I am OK with giving it a shot, but it does make me a bit less confident that nobody might be working around this in weird ways.
> 
> I guess we merge it ping the mailing list?

Sounds good.  You want me to send an email to the mailing list?

**Comment by seberg** [comment]
Happy if you do :).

**Comment by charris** [comment]
@Kai-Striega The suggestion has been made to move busday functionality to `numpy-financial` at some point. What are your thoughts about that suggestion? It might be a bit messy, as the code is in C and depends on `numpy.datetime`.

**Comment by stefanv** [comment]
> Counts the number of valid days between `begindates` and
`enddates`, not including the day of `enddates`.
>
> If ``enddates`` specifies a date value that is earlier than the
> corresponding ``begindates`` date value, the count will be negative.

The function calculates the delta for a half open bracket. Question is: what is that bracket when you have enddates < begindates.

For indexing, we always exclude the last:

```
In [13]: x[1:4]
Out[13]: array([1, 2, 3])

In [15]: x[4:1:-1]
Out[15]: array([4, 3, 2])
```

Current behavior:

```
In [16]: np.busday_count(np.datetime64('2023-03-12'), np.datetime64('2023-03-13'))
Out[16]: 0
```

Sunday to Monday, with Monday excluded. Makes sense. But:

```
In [17]: np.busday_count(np.datetime64('2023-03-13'), np.datetime64('2023-03-12'))
Out[17]: 0
```

Monday through Sunday, would have expected -1.

The current behavior (a) does not match the docstring and (b) is hard to predict, so would support fixing it: always exclude the enddates.

**Comment by Kai-Striega** [comment]
> @Kai-Striega The suggestion has been made to move busday functionality to numpy-financial at some point. What are your thoughts about that suggestion? It might be a bit messy, as the code is in C and depends on numpy.datetime.

Just a few thoughts:

* I believe the functionality would fit in with `numpy-financial`'s scope
* They *could* be useful for someone – but I can't think of a use case of the top of my head
* I don't have much experience working with / maintaining C, however, this seems like a good time to learn
* This would require a new release of `numpy-financial`. I've been planning this for a while anyway, it might be a good time to do one.

**Comment by eendebakpt** [comment]
A message was sent to numpy-discussion for comments: https://mail.python.org/archives/list/numpy-discussion@python.org/thread/MGTAMT5ELASBMVMCWU2IOPR5CQLD7CUP/

**Comment by seberg** [comment]
Let's give this a shot then, thanks.
